[ONNX] Fix for batchnorm training op mode

neginraoof opened the pull request:
Batchnorm should be exported in trainin/eval mode regardless of model's training mode.
This use-case is important for fine tuning models.

jiafatom reviewed (approved):
lgtm, please fix test errors.

BowenBao commented on torch/onnx/symbolic_opset12.py:
```diff
@@ -17,9 +17,9 @@ def einsum(g, equation, tensor_list):
 def dropout(g, input, p, train):
-    sym_help.assert_training_mode(train, "dropout")
+    sym_help.check_training_mode(train, "dropout")
     # in eval mode, dropout is non-op - if the node's train param is set to False, dropout is non-op
```
update comment?

neginraoof replied:
Do you mean the comment in line 21? I think this is not changed, since it is pointing to the training param for this node (op).

shubhambhokare1 reviewed (approved):
LGTM, thanks

fatcat-z commented on test/onnx/test_pytorch_onnx_onnxruntime.py:
```diff
@@ -7372,11 +7480,63 @@ def forward(self, x):
-         zip(ort_outs1, ort_outs2)]
+        model.eval()
+        pt_outs2 = model(x)
+        [np.testing.assert_allclose(ort_out2, pt_out2.detach(), atol=1e-4, rtol=0.001) for ort_out2, pt_out2 in
```
Change the format of rtol=0.001 to rtol=1e-3?

In addition, for atol, our normal testing uses 'atol=1e-7'. Should we keep the same one here?

fatcat-z commented on test/onnx/test_pytorch_onnx_onnxruntime.py:
```diff
@@ -7372,11 +7480,63 @@ def forward(self, x):
                                     training=torch.onnx.TrainingMode.TRAINING)
         ort_outs1 = run_ort(ort_sess1, input=(x,))
+        pt_outs1 = model(x)
+        [np.testing.assert_allclose(ort_out1, pt_out1.detach(), atol=1e-4, rtol=0.001) for ort_out1, pt_out1 in
```
Change the format of rtol=0.001 to rtol=1e-3?

In addition, for atol, our normal testing uses 'atol=1e-7'. Should we keep the same one here?

fatcat-z commented on torch/onnx/symbolic_helper.py:
```diff
@@ -697,19 +697,23 @@ def _avgpool_helper(tuple_fn, padding, kernel_size, stride, divisor_override, na
                       " mode, but operator " + op_name + " is set to " +
-                      op_mode + " mode. The model will be exported in " +
-                      training_mode + ", as specified by the export mode.")
+                      op_mode + " mode. The operatos will be exported in " +
```
Typo: operatos -> operator

shubhambhokare1 reviewed (commented):
Overall looks great and great test coverage, some minor comments (including Jay's)

BowenBao commented on test/onnx/test_operators.py:
```diff
@@ -681,7 +681,7 @@ def test_dropout_training(self):
     def test_dropout_opset12(self):
         x = torch.randn(3, 4, requires_grad=True)
-        self.assertONNX(lambda x: torch.max(functional.dropout(x)), x, opset_version=12)
+        self.assertONNX(lambda x: torch.max(functional.dropout(x, training=False)), x, opset_version=12)
```
what is this change for?

neginraoof replied:
training is True by default

BowenBao commented on torch/onnx/symbolic_helper.py:
```diff
@@ -700,19 +700,23 @@ def _avgpool_helper(tuple_fn, padding, kernel_size, stride, divisor_override, na
+        # This is to support use-cases of fixing certain layer weights
+        # in training.
         warnings.warn("ONNX export mode is set to " + training_mode +
                       " mode, but operator " + op_name + " is set to " +
```
Not to be done in this PR, we should update node naming to be consistent throughout IR graph and exported model, such that we can print which particular node this is in warning messages.

BowenBao commented on torch/onnx/symbolic_helper.py:
```diff
@@ -700,19 +700,23 @@ def _avgpool_helper(tuple_fn, padding, kernel_size, stride, divisor_override, na
-    op_mode = True if op_mode == 1 else False
-    if op_mode != _training_mode:
-        op_mode = "training " if op_mode else "inference"
+    op_train_mode = True if op_train_mode == 1 else False
```
Checking my understanding, how is the op_mode created for operators in the first place? 
I assume users could set modes for ops explicitly. And without it the op will take the model mode. Is that correct?

neginraoof replied:
Yes, see example in: test_batchnorm_eval_mode_train_layer

BowenBao commented on torch/csrc/jit/passes/onnx/eval_peephole.cpp:
```diff
@@ -46,14 +46,18 @@ static void fuseConvBatchNorm(Block* b, ValueToParamPairMap& valsToParamsMap) {
-      auto origconvNode = *it;
+      // Even when model is in eval mode, check if BatchNorm layer is in
+      // training mode
+      if (bnNode->outputs().size() != 1)
```
does batchnorm has a training attribute to check with, or is this only achievable by looking at output count?

neginraoof replied:
Starting from opset 14, it has an attribute.

BowenBao commented on test/onnx/test_pytorch_onnx_onnxruntime.py:
```diff
@@ -7592,24 +7676,50 @@ def forward(self, x):
+        [np.testing.assert_allclose(ort_out2, pt_out2.detach(), atol=1e-4, rtol=0.001) for ort_out2, pt_out2 in
+         zip(ort_outs2, [pt_outs2])]
+
+    @unittest.skip("Disable due to missing shape of Conv op output as BN input")
```
is this still active? why is it not resolved by shape inference?

BowenBao commented on test/onnx/test_pytorch_onnx_onnxruntime.py:
```diff
@@ -7456,42 +7456,126 @@ def test_batchnorm_training(self):
+        outs = model_export(x)
+        pytorch_out = [out.detach().numpy() for out in outs]
+        ort_sess = convert_to_onnx(model_export, input=(x,), opset_version=self.opset_version, example_outputs=outs,
+                                   training=torch.onnx.TrainingMode.TRAINING,
```
why calling convert_to_onnx and compare results explicitly? Would adding training flag to `run_test` enough to resolve this?

BowenBao commented on test/onnx/test_pytorch_onnx_onnxruntime.py:
```diff
@@ -7641,11 +7751,63 @@ def forward(self, x):
+        [np.testing.assert_allclose(ort_out2, pt_out2.detach(), atol=1e-4, rtol=0.001) for ort_out2, pt_out2 in
+         zip(ort_outs2, [pt_outs2])]
+
+    @unittest.skip("Disable due to missing shape of Conv op output as BN input")
```
ditto

BowenBao commented:
Looks good overall, have a few questions on some details. Also please check the CI failure.

neginraoof committed:
Update symbolic_opset12.py

neginraoof committed:
Merge

shubhambhokare1 commented on torch/onnx/symbolic_opset9.py:
```diff
@@ -1378,8 +1378,8 @@ def batch_norm(g, input, weight, bias, running_mean, running_var, training, mome
                momentum_f=1 - momentum,
-               outputs=1 if not sym_help._training_mode else 5)
-    if not sym_help._training_mode:
+               outputs=1 if not training else 5)
```
Might need to replicate this for opset 14 BN

neginraoof replied:
Well this PR does not have opset 14 updates yet. I'll have to do that after rebase.

shubhambhokare1 commented on test/onnx/test_utility_funs.py:
```diff
@@ -917,21 +921,6 @@ def forward(self, x, y):
                                (TestCase,),
                                dict(TestUtilityFuns.__dict__, opset_version=13))
 
-# opset 11 tests
```
Have removed this opset 14, might cause conflicts

shubhambhokare1 reviewed (commented):
Overall looks good, could you look at batchnorm_helper added in opset 14 PR

neginraoof commented:
> Overall looks good, could you look at batchnorm_helper added in opset 14 PR

Well this PR does not have opset 14 updates yet. I'll have to do that after rebase.

fatcat-z commented on torch/csrc/jit/passes/onnx/eval_peephole.cpp:
```diff
@@ -47,14 +47,17 @@ static void fuseConvBatchNorm(Block* b, ValueToParamPairMap& valsToParamsMap) {
     if (it->kind() == onnx::Conv) {
-      if (it->output()->uses().size() != 1) {
+      auto oldConv = *it;
+      auto bnNode = oldConv->outputs().at(0)->uses()[0].user;
```
Does this change mean that we expect the Conv op will have more than 1 output, not like before anymore?
If so, how can we assure that the first output is BatchNorm? Should we go through all of existing outputs to find out if BatchNorm op exists?

neginraoof replied:
No, I don't expect Conv node to have more than one outputs.
The use of "at(i)" would throw a runtime error instead of a seg-fault in case of an issue when we mess up the nodes.

BowenBao commented on torch/csrc/jit/passes/onnx/eval_peephole.cpp:
```diff
@@ -47,14 +47,17 @@ static void fuseConvBatchNorm(Block* b, ValueToParamPairMap& valsToParamsMap) {
       fuseConvBatchNorm(child_block, valsToParamsMap);
     }
     if (it->kind() == onnx::Conv) {
-      if (it->output()->uses().size() != 1) {
```
We should need to check bn is the only user of this conv right?

neginraoof replied:
Just added this back. Thanks.

neginraoof committed:
Fix for comment

BowenBao reviewed (approved):
LGTM, thanks!

neginraoof committed:
Rebase

# Conflicts:
#	torch/csrc/jit/passes/onnx/eval_peephole.cpp

# Conflicts:
#	test/onnx/test_utility_funs.py
#	torch/onnx/symbolic_opset9.py

neginraoof committed:
Update test.sh
# Conflicts:
#	.jenkins/caffe2/test.sh

neginraoof committed:
Fix utility tests

# Conflicts:
#	test/onnx/test_pytorch_onnx_onnxruntime.py
#	test/onnx/test_utility_funs.py

neginraoof committed:
Enable BN tests

neginraoof committed:
Fix for test

neginraoof committed:
Update test_pytorch_onnx_onnxruntime.py

neginraoof committed:
Update test_pytorch_onnx_onnxruntime.py

neginraoof committed:
Update test_utility_funs.py

neginraoof committed:
Update test_pytorch_onnx_onnxruntime.py
